server/api.go

```diff
@@ -169,133 +171,66 @@ func (server *GripServer) addEdge(ctx context.Context, elem *gripql.GraphElement
-				graph, err := server.db.Graph(vBatch.graph)
-				if err != nil {
-					log.WithFields(log.Fields{"error": err}).Error("BulkAdd: graph connection error")
-					continue
-				}
-				err = graph.AddVertex(vBatch.vertices)
-				if err != nil {
-					log.WithFields(log.Fields{"error": err}).Error("BulkAdd: add vertex error")
-				}
-			}
-		}
-		closeChan <- true
-	}()
-
-	go func() {
-		for eBatch := range edgeBatchChan {
-			if isSchema(eBatch.graph) {
-				err := "cannot add edge to schema graph"
-				log.WithFields(log.Fields{"error": err}).Error("BulkAdd: add edge error")
-				continue
-			}
-			if len(eBatch.edges) > 0 && eBatch.graph != "" {
-				graph, err := server.db.Graph(eBatch.graph)
-				if err != nil {
-					log.WithFields(log.Fields{"error": err}).Error("BulkAdd: graph connection error")
-					continue
-				}
-				err = graph.AddEdge(eBatch.edges)
-				if err != nil {
-					log.WithFields(log.Fields{"error": err}).Error("BulkAdd: add edge error")
-				}
-			}
-		}
-		closeChan <- true
-	}()
-
-	vertexBatch := newGraphElementArray("", vertexBatchSize, 0)
-	edgeBatch := newGraphElementArray("", 0, edgeBatchSize)
+	wg := &sync.WaitGroup{}
 	var loopErr error
 	for loopErr == nil {
 		element, err := stream.Recv()
 		if err == io.EOF {
-			if vertCount != 0 {
-				log.Debugf("%d vertices streamed", vertCount)
-			}
-			if edgeCount != 0 {
-				log.Debugf("%d edges streamed", edgeCount)
-			}
-			vertexBatchChan <- vertexBatch
-			edgeBatchChan <- edgeBatch
 			loopErr = err
 		} else if err != nil {
 			log.WithFields(log.Fields{"error": err}).Error("BulkAdd: streaming error")
 			loopErr = err
 		} else {
-			if element.Vertex != nil {
-				if vertexBatch.graph != element.Graph || len(vertexBatch.vertices) >= vertexBatchSize {
-					vertexBatchChan <- vertexBatch
-					vertexBatch = newGraphElementArray(element.Graph, vertexBatchSize, 0)
-				}
-				vertex := element.Vertex
-				err := vertex.Validate()
-				if err != nil {
-					return fmt.Errorf("vertex validation failed: %v", err)
-				}
-				vertexBatch.vertices = append(vertexBatch.vertices, vertex)
-				vertCount++
-			} else if element.Edge != nil {
-				if edgeBatch.graph != element.Graph || len(edgeBatch.edges) >= edgeBatchSize {
-					edgeBatchChan <- edgeBatch
-					edgeBatch = newGraphElementArray(element.Graph, 0, edgeBatchSize)
+			if isSchema(element.Graph) {
+				err := "cannot add element to schema graph"
+				log.WithFields(log.Fields{"error": err}).Error("BulkAdd: add element error")
+			} else {
+				if element.Graph != graphName {
+					if graphName != "" {
+						close(elementStream)
+					}
+					graph, err := server.db.Graph(element.Graph)
+					if err != nil {
+						log.WithFields(log.Fields{"error": err}).Error("BulkAdd: graph not found")
+						loopErr = err
+					}
+					elementStream = make(chan *gripql.GraphElement, 100)
+					wg.Add(1)
+					go func() {
+						log.Printf("Streaming to %s", element.Graph)
+						graph.BulkAdd(elementStream)
```

Comment:
BulkAdd error needs to be handled.

Reply:
The error is copied out now. Note, the current BulkAdd behavior is to cache an error message, keep on processing messages and return the error at the end